clean_air/util/crs.py

```diff
@@ -0,0 +1,250 @@
+
+    # Split the parameters into those for the globe and those for the crs
+    # Note: the sorting is specifically to guarantee that we handle lat_1
+    # before lat_2, because they need to be put into an array in that order
+    unrecognised = []
+    for key, val in sorted(params.items()):
+        # Handle "flags" - parameters that have no meaningful "value" so
+        # were given a value of None
+        if key == "south":
+            val = True
+
+        if key in _GLOBE_PARAMS:
+            globe_params[_GLOBE_PARAMS[key]] = val
+        elif key in _CRS_PARAMS:
+            name = _CRS_PARAMS[key]
+            if name.startswith("["):
+                # Force into an array if needed
+                name = name.strip("[]")
+                crs_params.setdefault(name, []).append(val)
+            else:
+                crs_params[name] = val
+        elif key not in ("no_defs", "wktext", "type", "units", "to_meter"):
+            unrecognised.append((key, val))
+
+    if unrecognised:
+        warnings.warn(f"Some parameters were not handled: {unrecognised}")
+
+    # Special handling for scale factors
+    # Cartopy refuses to accept both "scale_factor" and "latitude_true_scale",
+    # even if they match the defaults that proj4 defines.  Try to avoid it
+    # raising this error by removing these defaults.
+    if crs_params.get("latitude_true_scale") == 0:
+        crs_params.pop("latitude_true_scale")
+    if crs_params.get("scale_factor") == 1:
+        crs_params.pop("scale_factor")
+
+    return constructor(**crs_params, globe=ccrs.Globe(**globe_params))
+
+
+def _transformer_cartopy(source, target):
+    def transform(xs, ys):
+        tfpoints = target.transform_points(source, np.array(xs), np.array(ys))
+        # Cartopy gave us an array of shape (n, 3), but shapely expects output
+        # of the same type as the input, ie two lists of length n.
+        # A numpy array of shape (2, n) counts as two lists, so just need to
+        # drop the z coordinates and transpose.
+        return tfpoints[:, 0:2].T
+
+    return transform
+
+def _transformer_pyproj(source, target):
+    transformer = pyproj.Transformer.from_crs(source, target, always_xy=True)
+    return transformer.transform
+
+def transform_shape(shape, source, target):
+    # Determine an appropriate transformation function based on type
+    # If the CRSs are not compatible types, assume it is slightly more helpful
+    # to match the target, so convert the source
+    if isinstance(source, ccrs.CRS):
+        if isinstance(target, ccrs.CRS):
+            transformer = _transformer_cartopy(source, target)
+        else:
+            source = as_pyproj_crs(source)
+            transformer = _transformer_pyproj(source, target)
+    else:
+        if isinstance(target, ccrs.CRS):
+            source = as_cartopy_crs(source)
+            transformer = _transformer_cartopy(source, target)
+        else:
+            transformer = _transformer_pyproj(source, target)
```

Comment:
I think that link is just a list of the ways of describing CRSs, in general, and saying that geopandas can handle any of them because pyproj can. There may well be other python packages that provide a CRS class to worry about anyway, that's just not a list of them. Pyproj and cartopy gets us a long way I think.